plutus-tx-plugin/test/Blueprint/Tests.hs

```diff
@@ -10,92 +11,63 @@
 
 module Blueprint.Tests where
 
-import PlutusTx.Blueprint
-import Prelude
-
+import Blueprint.Tests.Lib (goldenJson)
 import Blueprint.Tests.Lib qualified as Fixture
-import Control.Monad.Reader (asks)
 import Data.Set qualified as Set
-import Data.Void (Void)
+import PlutusTx.Blueprint
 import PlutusTx.Blueprint.Purpose qualified as Purpose
-import PlutusTx.Builtins (BuiltinByteString, BuiltinData)
-import System.FilePath ((</>))
-import Test.Tasty (TestName)
+import Prelude
 import Test.Tasty.Extras (TestNested, testNested)
-import Test.Tasty.Golden (goldenVsFile)
 
 goldenTests :: TestNested
-goldenTests = testNested "Blueprint" [goldenBlueprint "Acme" contractBlueprint]
+goldenTests = testNested "Blueprint" [goldenJson "Acme" (`writeBlueprint` contractBlueprint)]
 
-goldenBlueprint :: TestName -> ContractBlueprint types -> TestNested
-goldenBlueprint name blueprint = do
-  goldenPath <- asks $ foldr (</>) name
-  let actual = goldenPath ++ ".actual.json"
-  let golden = goldenPath ++ ".golden.json"
-  pure $ goldenVsFile name golden actual (writeBlueprint actual blueprint)
-
-{- | All the data types exposed (directly or indirectly) by the type signature of the validator
-This type level list is used to:
-1. derive the schema definitions for the contract.
-2. make "safe" references to the [derived] schema definitions.
--}
-type ValidatorTypes =
-  [ Fixture.Datum
-  , Fixture.DatumPayload
-  , Fixture.Params
-  , Fixture.Redeemer
-  , Fixture.Bytes Void
-  , ()
-  , Bool
-  , Integer
-  , BuiltinData
-  , BuiltinByteString
-  ]
-
-contractBlueprint :: ContractBlueprint ValidatorTypes
+-- | All the data types exposed (directly or indirectly) by the type signature of the validator
+-- This type level list is used to:
+-- 1. derive the schema definitions for the contract.
+-- 2. make "safe" references to the [derived] schema definitions.
+contractBlueprint :: Blueprint [Fixture.Params, Fixture.Redeemer, Fixture.Datum]
```

Comment:
This is better than before but there's still repetition: you are specifying the param, redeemer and datum types here, but then have to repeat them in `parameterSchema = definitionRef @Fixture.Params`, `argumentSchema = definitionRef @Fixture.Redeemer` and `argumentSchema = definitionRef @Fixture.Datum`. It should be possible to require writing the only once.

Reply:
The thing is that one contract blueprint can describe more than one validator. Imagine the following contract, for example:
```haskell
data V1Params = MkV1Params Integer
type V1Redeemer = ()
type V1Datum = Integer
data V2Params = MkV2Params Bool
data V2Redeemer = RedeemThis | RedeemThat
data V2Datum = Bool

validator1 :: V1Params -> V1Redeemer -> V1Datum -> ScriptContext -> Bool
validator2 :: V2Params -> V2Redeemer -> V2Datum -> ScriptContext -> Bool
```

Current implementation allows to define

```haskell
contractBlueprint :: Blueprint [V1Params, V1Redeemer, V1Datum, V2Params, V2Redeemer, V2Datum]
```
and then reference corresponding schema from a validator blueprint, e.g. first `ValidatorBlueprint` references `V1*` and second one references `V2*`.

With such case I don't see how it could be possible to write top-level types (Param/Redeemer/Datum) once, as there will be a need to specify which top level type belongs to which validator. 

I can think of something along these lines:

```haskell
multipleValidatorContract :: Blueprint 
    [ [V1Params, V1Redeemer, V1Datum] -- first validator
    , [V2Params, V2Redeemer, V2Datum] -- second validator
    ]
    
type Blueprint1 p r d = Blueprint '[[p, r, d]]
singleValidatorContract :: Blueprint1 MyParams MyRedeemer MyDatum    
```
but such solution is more complex. 

@zliu41 @michaelpj wdyt, which design is better?

Reply:
Is it possible to make it so that writing `definitionRef @V2Params` etc. is sufficient to register `V2Params` as a used type, and avoid needing a type-level list at the top? You can do this separately.

Reply:
This is a good idea: a definition is registered by mere fact of referencing it - this is another way of making sure there are no broken references!

Unfortunately I don't yet know how an implementation of such feature could look like. What comes to mind is that every definition reference contributes a constraint, something like:
```haskell
definitionRef :: forall t ts. HasDefinition t ts => Schema ts


```
and then all such constraints are accumulated at the top level

```haskell
( HasDefinition Params ts
, HasDefinition Datum ts
, HasDefinition Redeemer ts
, HasDefinition Integer ts
, HasDefinition Bool ts
) => ...
```
and this set of constraints then drives the `deriveSchemaDefinitions` ...

I'd need to give this approach a "spike" try as I am not sure it can work.

Reply:
>  I'm suggesting to replace deriveSchemaDefinitions with something like [ deriveDefinition @Foo, deriveDefinition @Bar ].

With this approach I don't see how to make references that lack a definition entry fail in compile time.

Reply:
I found a way to make type inference work the way @michaelpj has described:
1. User lists top-level contract types;
2. Users's type is un-rolled to include all nested types;
3. Schema definitions are derived from the un-rolled list;
4. Schema references are checked against the un-rolled list;
5. The `referencedTypes` type variable in the `ContractBlueprint referencedTypes` which holds un-rolled list is existentially hidden using a wrapper data type;